### ISSUE TITLE
Crash due to debugger command line length

### ISSUE THREAD

**Issue by xotmatrix** [AppleWin/AppleWin#1204]
When typing a command in the debugger, if it's length + prompt + cursor exceeds the window's width (nominally 80 characters), the command line will appear glitched around the prompt (1). At this point, backspacing to shorten the command does not appear to remove characters and the prompt remains glitched (2). At this point, entering the command glitches the output with an overlapping response or improper scroll (3). Entering a command longer than will fit the window causes a silent crash.

1.
![2023-03-29_20-46-39](https://user-images.githubusercontent.com/14209329/228699798-f30e5d54-5479-4277-93e4-4fc56de411ef.png)
2.
![2023-03-29_20-47-24](https://user-images.githubusercontent.com/14209329/228699805-4bccf1e8-56d2-4dc3-8acc-0fd820c9b69d.png)
3.
![2023-03-29_20-47-58](https://user-images.githubusercontent.com/14209329/228699815-296eb890-f6ff-4186-8942-f624d831b141.png)

**Comment by Michaelangel007** [comment]
xotmatrix Thanks for filing this bug.  Long command lines are a known issue but it is good to have it officially be tracked now.  (I've been holding off for years to fix proper command line editing and length.)

At the very least the debugger shouldn't crash.

**Comment by Michaelangel007** [comment]
@xotmatrix What's the maximum input length you want/need?

**Comment by xotmatrix** [comment]
I don't have a particular need. This problem presented itself in working on a debugger with a CLI for my own emulator. In my initial CLI, I had a dedicated input line, like AppleWin. It had an indefinite length and when it reached and exceeded 80 characters, it would show the last 79 characters of the command followed by the cursor, creating a scrolling effect as one typed. That's what I expected AppleWin to do, so I tested it and here we are.

**Comment by Michaelangel007** [comment]
**Analysis:**

Off-by-one bug is in _Debugger_Console.cpp_ `ConsoleInputChar()`:

```c
bool ConsoleInputChar ( char ch )
{
	if (g_nConsoleInputChars < g_nConsoleDisplayWidth) // bug? include prompt?
```

**Future Work:**

The expression ` (g_nConsoleDisplayWidth-1)` can be factored out. It also shows up in _Debug.cpp_ `DebuggerINputConsoleChar()`:

```c
	if (g_nConsoleInputChars > (g_nConsoleDisplayWidth-1))
		return;
```

To support an input line > 78 chars a few other changes need to be made:

* In _Debugger_Display.c_ `DrawConsoleInput()` needs to be updated to show a horizontal scrolling window,
* In _Debug.cpp_ `DebuggerProcessCommand()` calls ConsoleDisplayPush() but this does NOT split long strings up into multiple pushes.

```c
`	if (bEchoConsoleInput)
		ConsoleDisplayPush( ConsoleInputPeek() );
```

There are also issues with `Backspace` for long lines.  _Debugger_Console.cpp_ `ConsoleInputBackSpace()` replaces EOL with spaces.  I replaced this with `0` but more testing is needed.

**Comment by tomcw** [comment]
@xot - Fixed in [AppleWin 1.30.15.0](https://github.com/AppleWin/AppleWin/releases/tag/v1.30.15.0).
